// @microsoft/vscode-container-client/lib/utils/AccumulatorStream.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See LICENSE in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccumulatorStream = void 0;
const stream = require("stream");
/**
 * Accumulates stream content in memory, which can then be obtained with
 * `getBytes()` or `getString()`. Both are async and will resolve only
 * once the stream has ended.
 */
class AccumulatorStream extends stream.Writable {
    /**
     * Creates an {@link AccumulatorStream}
     * @param options Same as {@link stream.WritableOptions}, except `write` or `writev` cannot be supplied.
     */
    constructor(options) {
        super({
            ...options,
            write: (chunk, encoding, callback) => {
                this.chunks.push(chunk);
                callback();
            },
        });
        this.chunks = [];
        this.streamEndPromise = new Promise((resolve, reject) => {
            this.on('close', () => {
                resolve();
            });
            this.on('error', (err) => {
                reject(err);
            });
        });
    }
    /**
     * Gets the full stream content
     * @returns The full stream content in a {@link Buffer}
     */
    async getBytes() {
        await this.streamEndPromise;
        return Buffer.concat(this.chunks);
    }
    /**
     * Gets the full stream content
     * @returns The full stream content in a string
     */
    async getString(encoding = 'utf-8') {
        const rawString = (await this.getBytes()).toString(encoding);
        // Remove non-printing control characters and trailing newlines
        // eslint-disable-next-line no-control-regex
        return rawString.replace(/[\x00-\x09\x0B-\x0C\x0E-\x1F]|\r?\n$/g, '');
    }
}
exports.AccumulatorStream = AccumulatorStream;
