// @microsoft/vscode-container-client/lib/clients/DockerClientBase/DockerInspectVolumeRecord.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See LICENSE in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.normalizeDockerInspectVolumeRecord = exports.isDockerInspectVolumeRecord = void 0;
const dayjs_1 = require("../../utils/dayjs");
// TODO: Actually test properties
function isDockerInspectVolumeRecord(maybeVolume) {
    return true;
}
exports.isDockerInspectVolumeRecord = isDockerInspectVolumeRecord;
function normalizeDockerInspectVolumeRecord(volume) {
    const createdAt = dayjs_1.dayjs.utc(volume.CreatedAt);
    // Return the normalized InspectVolumesItem record
    return {
        name: volume.Name,
        driver: volume.Driver,
        mountpoint: volume.Mountpoint,
        scope: volume.Scope,
        labels: volume.Labels,
        options: volume.Options,
        createdAt: createdAt.toDate(),
        raw: JSON.stringify(volume),
    };
}
exports.normalizeDockerInspectVolumeRecord = normalizeDockerInspectVolumeRecord;
