// @microsoft/vscode-container-client/lib/clients/DockerClientBase/parseListFilesCommandOutput.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See LICENSE in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseListFilesCommandWindowsOutput = exports.parseListFilesCommandLinuxOutput = void 0;
const path = require("path");
const vscode = require("vscode");
const dayjs_1 = require("../../utils/dayjs");
const DateFormats = [
    'MMM D HH:mm',
    'MMM D YYYY',
    'MM/DD/YYYY hh:mm A', // Windows format
];
const SupportedFileTypes = [vscode.FileType.Directory, vscode.FileType.File];
function parseListFilesCommandLinuxOutput(options, output) {
    const regex = /^(?<type>[0-9a-fA-F]+)\s+(?<links>\d+)\s+(?<group>\d+)\s+(?<owner>\d+)\s+(?<size>\d+)\s+(?<atime>\d+)\s+(?<mtime>\d+)\s+(?<ctime>\d+)\s+(?<name>.*)$/gm;
    const items = new Array();
    for (const match of output.matchAll(regex)) {
        /* eslint-disable @typescript-eslint/no-non-null-assertion */
        const name = path.basename(match.groups.name);
        const { mode, fileType } = parseLinuxType(match.groups.type);
        const size = Number.parseInt(match.groups.size, 10);
        const mtime = dayjs_1.dayjs.unix(Number.parseInt(match.groups.mtime, 10)).valueOf();
        const ctime = dayjs_1.dayjs.unix(Number.parseInt(match.groups.ctime, 10)).valueOf();
        const atime = dayjs_1.dayjs.unix(Number.parseInt(match.groups.atime, 10)).valueOf();
        /* eslint-enable @typescript-eslint/no-non-null-assertion */
        // Ignore relative directory items...
        if (fileType === vscode.FileType.Directory && (name === '.' || name === '..')) {
            continue;
        }
        // Ignore everything other than directories and plain files
        if (!SupportedFileTypes.includes(fileType)) {
            continue;
        }
        items.push({
            name,
            path: path.posix.join(options.path, name),
            type: fileType,
            mode,
            ctime,
            mtime,
            atime,
            size,
        });
    }
    return items;
}
exports.parseListFilesCommandLinuxOutput = parseListFilesCommandLinuxOutput;
function parseListFilesCommandWindowsOutput(options, output) {
    const regex = /^(?<mtime>(?<date>\d{1,2}(\/|\.)\d{1,2}(\/|\.)\d{4})\s+(?<time>\d{1,2}:\d{1,2}( (AM|PM))?))\s+((?<type><DIR>|<SYMLINKD>)|(?<size>\d+))\s+(?<name>.*)$/gm;
    const items = new Array();
    for (const match of output.matchAll(regex)) {
        /* eslint-disable @typescript-eslint/no-non-null-assertion */
        const name = match.groups.name;
        const fileType = parseWindowsType(match.groups.type);
        const size = fileType === vscode.FileType.Directory ? 0 : Number.parseInt(match.groups.size, 10);
        const mtime = (0, dayjs_1.dayjs)(match.groups.mtime, DateFormats).valueOf();
        /* eslint-enable @typescript-eslint/no-non-null-assertion */
        // Ignore relative directory items...
        if (fileType === vscode.FileType.Directory && (name === '.' || name === '..')) {
            continue;
        }
        // Ignore everything other than directories and plain files
        if (!SupportedFileTypes.includes(fileType)) {
            return items;
        }
        items.push({
            name,
            path: path.win32.join(options.path, name),
            type: fileType,
            mtime,
            size,
        });
    }
    return items;
}
exports.parseListFilesCommandWindowsOutput = parseListFilesCommandWindowsOutput;
function parseLinuxType(fullModeHex) {
    const fullMode = parseInt(fullModeHex, 16);
    const fileType = (fullMode & 0xf000) >> 12;
    const mode = fullMode & 0xfff;
    switch (fileType) {
        case 4:
            return { mode, fileType: vscode.FileType.Directory };
        case 8:
            return { mode, fileType: vscode.FileType.File };
        case 10:
            return { mode, fileType: vscode.FileType.SymbolicLink };
        default:
            return { mode, fileType: vscode.FileType.Unknown };
    }
}
function parseWindowsType(type) {
    switch (type === null || type === void 0 ? void 0 : type.toUpperCase()) {
        case '<DIR>':
            return vscode.FileType.Directory;
        case '':
        case undefined:
            // Blank or undefined type is a file
            return vscode.FileType.File;
        case '<SYMLINKD>':
        case '<SYMLINK>':
            return vscode.FileType.SymbolicLink;
        default:
            return vscode.FileType.Unknown;
    }
}
