// @microsoft/vscode-container-client/lib/clients/PodmanClient/PodmanInspectNetworkRecord.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See LICENSE in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.normalizePodmanInspectNetworkRecord = exports.isPodmanInspectNetworkRecord = void 0;
function isPodmanInspectNetworkRecord(maybeNetwork) {
    const network = maybeNetwork;
    if (!network || typeof network !== 'object') {
        return false;
    }
    if (typeof network.name !== 'string') {
        return false;
    }
    if (network.labels && typeof network.labels !== 'object') {
        return false;
    }
    return true;
}
exports.isPodmanInspectNetworkRecord = isPodmanInspectNetworkRecord;
function normalizePodmanInspectNetworkRecord(network) {
    return {
        name: network.name,
        id: network.id,
        driver: network.driver,
        createdAt: network.created ? new Date(network.created) : undefined,
        internal: network.internal,
        ipv6: network.ipv6_enabled,
        labels: network.labels || {},
        scope: undefined,
        attachable: undefined,
        ingress: undefined,
        ipam: undefined,
        raw: JSON.stringify(network),
    };
}
exports.normalizePodmanInspectNetworkRecord = normalizePodmanInspectNetworkRecord;
