// @microsoft/vscode-container-client/lib/clients/PodmanClient/PodmanInspectVolumeRecord.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See LICENSE in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.normalizePodmanInspectVolumeRecord = exports.isPodmanInspectVolumeRecord = void 0;
function isPodmanInspectVolumeRecord(maybeVolume) {
    const volume = maybeVolume;
    if (!volume || typeof volume !== 'object') {
        return false;
    }
    if (typeof volume.Name !== 'string') {
        return false;
    }
    if (typeof volume.Driver !== 'string') {
        return false;
    }
    if (typeof volume.Mountpoint !== 'string') {
        return false;
    }
    if (typeof volume.CreatedAt !== 'string') {
        return false;
    }
    if (volume.Labels && typeof volume.Labels !== 'object') {
        return false;
    }
    if (typeof volume.Scope !== 'string') {
        return false;
    }
    if (volume.Options && typeof volume.Options !== 'object') {
        return false;
    }
    return true;
}
exports.isPodmanInspectVolumeRecord = isPodmanInspectVolumeRecord;
function normalizePodmanInspectVolumeRecord(volume) {
    return {
        name: volume.Name,
        driver: volume.Driver,
        mountpoint: volume.Mountpoint,
        createdAt: new Date(volume.CreatedAt),
        labels: volume.Labels || {},
        scope: volume.Scope,
        options: volume.Options || {},
        raw: JSON.stringify(volume),
    };
}
exports.normalizePodmanInspectVolumeRecord = normalizePodmanInspectVolumeRecord;
