// @microsoft/vscode-container-client/lib/clients/ConfigurableClient.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See LICENSE in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _ConfigurableClient_commandName, _ConfigurableClient_displayName, _ConfigurableClient_description;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfigurableClient = void 0;
class ConfigurableClient {
    constructor(id, commandName, displayName, description) {
        this.id = id;
        _ConfigurableClient_commandName.set(this, void 0);
        _ConfigurableClient_displayName.set(this, void 0);
        _ConfigurableClient_description.set(this, void 0);
        __classPrivateFieldSet(this, _ConfigurableClient_commandName, commandName, "f");
        __classPrivateFieldSet(this, _ConfigurableClient_displayName, displayName, "f");
        __classPrivateFieldSet(this, _ConfigurableClient_description, description, "f");
    }
    get commandName() {
        return __classPrivateFieldGet(this, _ConfigurableClient_commandName, "f");
    }
    set commandName(value) {
        __classPrivateFieldSet(this, _ConfigurableClient_commandName, value, "f");
    }
    get displayName() {
        return __classPrivateFieldGet(this, _ConfigurableClient_displayName, "f");
    }
    set displayName(value) {
        __classPrivateFieldSet(this, _ConfigurableClient_displayName, value, "f");
    }
    get description() {
        return __classPrivateFieldGet(this, _ConfigurableClient_description, "f");
    }
    set description(value) {
        __classPrivateFieldSet(this, _ConfigurableClient_description, value, "f");
    }
}
exports.ConfigurableClient = ConfigurableClient;
_ConfigurableClient_commandName = new WeakMap(), _ConfigurableClient_displayName = new WeakMap(), _ConfigurableClient_description = new WeakMap();
